### PR TITLE
Distinct colors editorial

### PR DESCRIPTION
**Pull request by timothygao8710**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

File: solutions/cses-1139.mdx [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+using namespace std;
+#define pb push_back
+
+struct BIT {
+		vector<int> bit;
+		int n;
+		BIT(int n) {
+				this->n = n + 1;
+				bit.resize(n+1);
+		}
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:43]:
```suggestion
		BIT(int n) : n(n + 1), bit(n + 1) { }
```

File: solutions/cses-1139.mdx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		int sum(int r) {
+				r++;
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:45]: my man using two tabs instead of 1
what a madlad

File: solutions/cses-1139.mdx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+				int ret = 0;
+				while (r > 0) {
+						ret += bit[r];
+						r -= r & -r;
+				}
+				return ret;
+		}
+		void update(int idx, int v) {
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:53]: newline between diff functions plz

File: solutions/cses-1139.mdx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+				idx++;
+				while (idx < n) {
+						bit[idx] += v;
+						idx += idx & -idx;
+				}
+		}
+};
+
+const int MAXN = 2e5 + 1;
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:62]: optional, but i'd love it if you made this 1 indexed

File: solutions/cses-1139.mdx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	lend[u] = idx;
+	for (int n : adj[u]) {
+		if (n == par) continue;
+		dfs(n, u);
+	}
+	tour[idx] = u; idx++;
+}
+
+int main() {
+	ios_base::sync_with_stdio(false); cin.tie(0);
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:80]: remove if it runs fine w/o this

File: solutions/cses-1139.mdx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	for (int i = 0; i < N - 1; i++) {
+		cin >> u >> v;
+		adj[u].pb(v);
+		adj[v].pb(u);
+	}
+	dfs(1);
+	BIT bit(N);
+	map<int, int> last;
+	// This map stores the rightmost occurrence of each color
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:93]: put the comment on top of the variable it describes like a sane person

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cses-1139.mdx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+ 
+using namespace std;
+#define pb push_back
+
+struct BIT {
+		vector<int> bit;
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:38]: still two tabs here

Comment (timothygao8710) [reply solutions/cses-1139.mdx]: alright fixed

File: solutions/cses-1139.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/cses-1139.mdx:78]: sooo it tle's w/o this?

File: solutions/cses-1139.mdx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		while (idx < n) {
+			bit[idx] += v;
+			idx += idx & -idx;
+		}
+	}
+};
+
+const int MAXN = 2e5 + 1;
+
+// Euler Tour, color of each node, answer for each node, left endpoint of each node (as explained in editorial)
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:62]: too long i think

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/cses-1139.mdx [hunk 9]
```diff
@@ -59,7 +59,10 @@ struct BIT {
 
 const int MAXN = 2e5 + 1;
 
-// Euler Tour, color of each node, answer for each node, left endpoint of each node (as explained in editorial)
+//tour = Euler Tour
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:62]: space after // for inline comment

File: solutions/cses-1139.mdx [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+			idx += idx & -idx;
+		}
+	}
+};
+
+const int MAXN = 2e5 + 1;
+
+// tour = Euler Tour
+// color = color of each node
+// answer = answer for each node
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:64]: i am so sorry timothy
but can you
```cpp
/*
 * format this comment something
 * like this
 */

File: solutions/cses-1139.mdx [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+vector<int> adj[MAXN];
+int idx = 0;
+
+void dfs(int u, int par = 0) {
+	lend[u] = idx;
+	for (int n : adj[u]) {
+		if (n == par) continue;
+		dfs(n, u);
+	}
+	tour[idx] = u; idx++;
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:77]: statements on separate lines

File: solutions/cses-1139.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/cses-1139.mdx:81]: still no update

File: solutions/cses-1139.mdx [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+		// If we ever considered the color of tour[i]
+		if (last.count(color[tour[i]])) { 
+			// The last time the color of tour[i] appeared, we make that 0 again
+			bit.update(last[color[tour[i]]], -1); 
+		}
+		// We change the last occurrence of the color of tour[i]
+		last[color[tour[i]]] = i; 
+		// We reflect that change in our BIT
+		bit.update(i, 1); 
+		// The answer for tour[i]'s node is simply the sum of 1s in its contiguous subinterval in the Euler Tour
```
Comment (SansPapyrus683) [inline solutions/cses-1139.mdx:107]: too long?

**Comment by SansPapyrus683** [comment]
> when can this be approved? It seems entirely finished.

~~until i'm done tearing it apart lol~~
yeah it's almost done

**Review by nwatx — APPROVED** [review]

**Comment by nwatx** [comment]
@SansPapyrus683 make those changes please

**Comment by SansPapyrus683** [comment]
what now

**Comment by SansPapyrus683** [comment]
wdym

File: solutions/cses-1139.mdx (hunk 12 above)
Comment (timothygao8710) [reply solutions/cses-1139.mdx:107]: bro like its honestly fine but ok

Comment (timothygao8710) [reply solutions/cses-1139.mdx:107]: btw i think its better to keep the fastio line in there anyways to speed up runtime.

Comment (SansPapyrus683) [reply solutions/cses-1139.mdx]: but why? as long as it's fast enough the fastio isn't needed

**Comment by SansPapyrus683** [comment]
you know what this pr's been sitting here for too long, screw it

**Comment by dutinmeow** [comment]
that's not how it works

**Comment by SansPapyrus683** [comment]
dustin you're just salty that I don't merge your pro

**Pull request merged**